plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+import PlutusTx.IsData qualified as P
+import PlutusTx.Prelude hiding (null, toList)
+import PlutusTx.These
+
+import Prelude qualified as Haskell
+
+{- | An associative map implementation backed by `P.BuiltinData`.
+
+This map implementation has the following characteristics:
+
+  * The `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations are no-op.
+  * Other operations are slower than @PlutusTx.AssocMap.Map@, although equality
+    checks on keys can be faster due to `P.equalsData`.
+  * Many operations involve converting the keys and/or values to/from `P.BuiltinData`.
+
+Therefore this map implementation is likely a better choice than @PlutusTx.AssocMap.Map@
+if it is part of a data type defined using @asData@, and the key and value types
+have efficient `P.toBuiltinData` and `P.unsafeFromBuiltinData` operations (e.g., they
+are primitive types or types defined using @asData@).
+-}
+newtype Map k a = Map P.BuiltinData
+  deriving stock (Haskell.Eq, Haskell.Show)
+  deriving newtype (Eq)
+
+instance P.ToData (Map k a) where
+  {-# INLINEABLE toBuiltinData #-}
+  toBuiltinData (Map d) = d
+
+instance P.FromData (Map k a) where
+  fromBuiltinData = Just . Map
+
+instance P.UnsafeFromData (Map k a) where
+  unsafeFromBuiltinData = Map
+
+{-# INLINEABLE lookup #-}
+lookup :: forall k a. (P.ToData k, P.UnsafeFromData a) => k -> Map k a -> Maybe a
+lookup (P.toBuiltinData -> k) m = case lookup' k (toBuiltinList m) of
+  Just a  -> Just (P.unsafeFromBuiltinData a)
+  Nothing -> Nothing
+
+{-# INLINEABLE lookup' #-}
+lookup' ::
+  BuiltinData ->
+  BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+  Maybe BuiltinData
+lookup' k = go
+  where
+    go xs =
+      P.matchList
+        xs
+        Nothing
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then Just (BI.snd hd)
+                  else go tl
+        )
+
+{-# INLINEABLE member #-}
+member :: forall k a. (P.ToData k) => k -> Map k a -> Bool
+member (P.toBuiltinData -> k) m = member' k (toBuiltinList m)
+
+{-# INLINEABLE member' #-}
+member' :: BuiltinData -> BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+member' k = go
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then True
+                  else go tl
+        )
+
+{-# INLINEABLE insert #-}
+insert :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a -> Map k a
+insert (P.toBuiltinData -> k) (P.toBuiltinData -> a) m =
+  unsafeFromBuiltinList (go (toBuiltinList m))
+  where
+    go ::
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) ->
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        (BI.mkCons (BI.mkPairData k a) nil)
```

Comment:
I think that if the key's not already in the map then this always inserts the new entry at the end and has to duplicate all of the original list when it does `BI.mkCons hd (go tl)` a few lines below here.  Could we somehow just cons the new entry onto the front in that case? That would presumably be more memory-efficient.

Later: I thought you could avoid duplicating the original input by binding it to a variable at the start and consing onto that if you arrive at the end without finding the key, but you _do_ have to copy the front of the list if the key actually is present and you can't tell that it's safe not to  do that until you get to the end.  Hmmm.

Later still: I suppose you could scan the whole list first to see if your key's already there and do the cheap cons if it isn't, but then you're paying for a possible memory saving with the CPU cost of an extra traversal.

Reply:
> I suppose you could scan the whole list first to see if your key's already there and do the cheap cons if it isn't, but then you're paying for a possible memory saving with the CPU cost of an extra traversal.

I doubt doing two traversals for insertion is the more efficient implementation in any programming language (do people do that in any language?), and thus isn't the right thing to do in PLC either.

Reply:
I suppose in general you'd use some structure that lets you do efficient insertion, but trying to do something like that in PLC might have too much overhead. The trouble is that we're trying to maintain an invariant of every key only occurring once, and that's maybe a bad fit with lists.  I suppose we could do some experiments to see what the space/time tradeoff is like if we do two traversals. Maybe the space usage isn't too bad though: you do have to copy all of the nodes if the thing you're inserting isn't in the map, but it's a shallow copy where you just need to allocate new cons-cells that contain pointers to the contents without copying the contents themselves.

Reply:
> Maybe the space usage isn't too bad 

... although if you create a map with n entries by calling `insert` n times you'll allocate n(n+1)/2 cons cells in total.